From 5d9501f024f2f0ba0bbdb772bb4f514374e51399 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <ramya.gnanasekar@oss.qualcomm.com>
Date: Wed, 22 Jan 2025 00:36:23 +0530
Subject: [PATCH] QSDK: CP: wifi: ath12k: Implement handshake for ml peer id

Maxmimum ml peer id supported by firmware depends on the platform. Since
ml_peer_id is generated by host, firmware shares the information in MLO
setup complete event.

Add code changes to decode the information from TLV and use it while
allocating the peer id.

Fixes: I11d96d3c8fdb4(UPSTREAM:CP:wifi:ath12k: Change ml peer id
allocation logic)

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h | 1 +
 drivers/net/wireless/ath/ath12k/peer.c | 6 +++---
 drivers/net/wireless/ath/ath12k/wmi.c  | 1 +
 drivers/net/wireless/ath/ath12k/wmi.h  | 1 +
 4 files changed, 6 insertions(+), 3 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1323,6 +1323,7 @@ struct ath12k_hw {
 	bool in_recovery;
 	u32 max_ml_peers_supported;
 	u32 num_ml_peers;
+	u32 max_ml_peer_ids;
 	struct ath12k radio[0] __aligned(sizeof(void *));
 };
 
--- a/drivers/net/wireless/ath/ath12k/peer.c
+++ b/drivers/net/wireless/ath/ath12k/peer.c
@@ -1087,8 +1087,8 @@ static u16 ath12k_mac_alloc_ml_peer_id(s
 	lockdep_assert_held(&ah->conf_mutex);
 
 	ml_peer_id = ah->last_ml_peer_id;
-	for (i = 0;  i < ATH12K_MAX_MLO_PEERS; i++) {
-		ml_peer_id = (ml_peer_id + 1) % ATH12K_MAX_MLO_PEERS;
+	for (i = 0;  i < ah->max_ml_peer_ids; i++) {
+		ml_peer_id = (ml_peer_id + 1) % ah->max_ml_peer_ids;
 
 		if (!ml_peer_id)
 			continue;
@@ -1101,7 +1101,7 @@ static u16 ath12k_mac_alloc_ml_peer_id(s
 	}
 
 	ah->last_ml_peer_id = ml_peer_id;
-	if (i == ATH12K_MAX_MLO_PEERS)
+	if (i == ah->max_ml_peer_ids)
 		ml_peer_id = ATH12K_MLO_PEER_ID_INVALID;
 
 	ath12k_dbg(NULL, ATH12K_DBG_PEER, "Allocated ml_peer_id:%d", ml_peer_id);
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -920,6 +920,7 @@ skip_lookup:
 	}
 
 	ar->mlo_setup_status = ev->status;
+	ar->ah->max_ml_peer_ids = ev->max_ml_peer_ids ? ev->max_ml_peer_ids : ATH12K_MAX_MLO_PEERS;
 	if (ab->hw_params->hal_ops->hal_get_tqm_scratch_reg)
 		ab->hw_params->hal_ops->hal_get_tqm_scratch_reg(ab, &ar->delta_tqm);
 	complete(&ar->mlo_setup_done);
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -8863,6 +8863,7 @@ struct wmi_mlo_teardown_fixed_param {
 struct wmi_mlo_setup_complete_event_fixed_param {
 	u32 pdev_id;
 	u32 status;
+	u32 max_ml_peer_ids;
 } __packed;
 
 struct wmi_mlo_teardown_complete_fixed_param {
